Ensure competition names are not greater than 50 characters when visible

pingskills opened the pull request:
I realised that the change I made to ensure a competition name <= 32 characters when it's not visible removed the existing validation for when the competition is visible. So I added back the original validation and added a new test for it to ensure competition names are <= 50 characters when visible.

jonatanklosko commented on WcaOnRails/app/models/competition.rb:
```diff
@@ -39,9 +40,12 @@ class Competition < ActiveRecord::Base
   def competition_name_validations
-    if name.length > MAX_NAME_LENGTH && !self.showAtAll?
+    if name.length > MAX_NAME_LENGTH_NOT_VISIBLE && !self.showAtAll?
       errors.add(:name, "Competition name can't be longer than 32 characters")
```
I suggest to use 'must not' just to be consistent (see the other errors below).

jonatanklosko commented on WcaOnRails/app/models/competition.rb:
```diff
@@ -39,9 +40,12 @@ class Competition < ActiveRecord::Base
+      errors.add(:name, "Competition name can't be longer than 50 characters")
+    end
     if !ENDS_WITH_YEAR_RE.match(name)
       errors.add(:name, "The competition name must end in a year.")
```
I would remove 'The' to keep the errors similarity.

jonatanklosko commented on WcaOnRails/app/models/competition.rb:
```diff
@@ -39,9 +40,12 @@ class Competition < ActiveRecord::Base
       errors.add(:name, "Competition name can't be longer than 32 characters")
     end
+    if name.length > MAX_NAME_LENGTH_VISIBLE && self.showAtAll?
+      errors.add(:name, "Competition name can't be longer than 50 characters")
```
Same as above.

jfly commented:
Good catch, Jeff! Thanks for adding a test :) I have nothing to add on top of jonatan's comments.

timhabermaas commented:
Something which isn't specific to this PR: Using `errors.add(:name, "Competition name is too long")` is inconsistent with how the built-in validations add error messages (they skip the model name and attribute part). This leads to the following errors hash/messages:

```ruby
{:id=>["can't be blank", "is invalid"], :cellName=>["is invalid"], :name=>["The competition name must end in a year."]}

errors.full_messages
# => ["Id can't be blank", "Id is invalid", "Competition nickname is invalid", "Name The competition name must end in a year."]
```

I think we should use `errors.add(:name, "can't be ...")` consistently when adding field specific error messages, otherwise inlining errors becomes awkward.

Another point: Could these validations be added using `validate :name, length: { maximum: 50 }, if: :showAtAll?` / `validate :name, length: { maximum: 32 }, unless: :showAtAll?`?

pingskills commented:
Thanks for the feedback everyone. I changed the validations to use errors.add(:name, "must not be ....") as suggested by @timhabermaas 

Let me know if you think there are more improvements to be made.

The form now displays the errors as follows:
![image](https://cloud.githubusercontent.com/assets/4403168/14819729/cb5b221c-0c06-11e6-9fd9-ef20104085d4.png)

jonatanklosko commented:
I think that custom validations aren't necessary here.
```ruby
  # As @timhabermaas indicated above
  validates :name, length: { maximum: MAX_NAME_LENGTH_NOT_VISIBLE }, unless: :showAtAll?
  validates :name, length: { maximum: MAX_NAME_LENGTH_VISIBLE }, if: :showAtAll?
  # And I suggest this too
  validates :name, format: { with: ENDS_WITH_YEAR_RE, message: "must end with a year" }
```
Conciseness is cool =)

pingskills commented:
Nice solution. Ill update the code. I think the custom message on the format is a good improvement over the original validation too.

pingskills commented:
When I update the code to have the following validations:

  validates :name, length: { maximum: MAX_NAME_LENGTH_VISIBLE }, if: :showAtAll?
  validates :name, length: { maximum: MAX_NAME_LENGTH_NOT_VISIBLE }, unless: :showAtAll?
  validates :name, format: { with: ENDS_WITH_YEAR_RE, message: "must end with a year" }

The view displays a form with a maxlength of whichever length validation comes first. So if I put the 50 character limit first then the user can type up to 50 characters. And if I do it the other way then when the comp is visible the user can only enter 32 characters and not 50. So by putting the 50 limit first it works but it would be nicer if the maxlength was set based on showAtAll?

![image](https://cloud.githubusercontent.com/assets/4403168/14848917/73865e96-0cb3-11e6-973f-bf388530ec37.png)

@jonatanklosko is there a way I can have the maxlength for the name be set dependent on showAtAll? rather than on what is first in the code?

jonatanklosko commented:
Oh I see. So I would suggest two ways to solve this (I'm not sure whether the first works).
```ruby
# First
validates :name, length: { maximum: (showAtAll? ? MAX_NAME_LENGTH_VISIBLE : MAX_NAME_LENGTH_NOT_VISIBLE) }

# Second
# Remove the constants with name length and define such a metchod
def max_name_length
  showAtAll? ? 50 : 32
end
# Validation
validates :name, length: { maximum: max_name_length }```

pingskills commented:
Hi @jonatanklosko

I couldn't get that code to work? Do you have any other thoughts?

![image](https://cloud.githubusercontent.com/assets/4403168/14853547/582ed6ac-0ccf-11e6-88ef-03fb8b7e6402.png)

pingskills commented:
I can't help but think that the right answer here would be to have a consistent maximum length of either 32 or 50 and not have any conditional logic.

jonatanklosko commented:
Hmm so we can't use methods in this way (I hadn't possibility to check that before responding). But I definitely agree that there should be the really **constant** length. I've just discovered that using two other max name lengths leads to validation error when you make a comp visible, change the name to something longer than 32 characters and then switch back it's visibility (that's an unbelievable case but still shows the inaccuracy).

pingskills commented:
The other option was to use a date to ensure future comps names are not longer than 32 chars and allow older comp names to be up to 50. Then you wouldn't get the problem mentioned above when switching visibility. 

We'd still have a max length="50" though for the form.

jonatanklosko commented:
I thought about this too and consider it to be reasonable. The only one problem here is that we don't have the created_at column in competitions table.

pingskills commented:
Here are the options I see:

1) Update the 154 competition names that are longer than 32 characters and enforce the 32 character limit. Time consuming but the best option if the 32 character limit is a real requirement.

2) Put the limit back to 50 characters. Best option if the 32 character limit is only "nice to have".

3) Validate using the showAtAll? so competitions that aren't visible must have names not more than 32 characters and competitions that are visible must have names not more than 50 characters. Not a great option since the maxlength of the input will be 50 and when entering a competition it's not easy to tell when you are over the limit. Plus there can be issues if people are toggling the visible flag.

I like option 1 or 2. Thoughts?

pedrosino commented:
I'm probably gonna ask something dumb, but why would the limits be different if the competition is visible or not?

The way I see it, every competition will become visible, so the first limit is somewhat pointless.

I didn't follow the discussion very well since I was travelling, but if I had to choose, I'd say option 2.

jfly commented:
@pingskills, sorry I led you in the wrong direction here. How about we go back to our original 50 character hard limit, and then just add a simple bootstrap .alert-warning for when competition name is > 32 characters? One nice implementation that would allow us to preserve the tests you wrote would be to add a `warnings` method to Competitions that returns an array of warnings. Then we could use that method in both tests and in our view.

pingskills commented:
@jfly it's actually been helpful for me to learn more about validations and test specs :)

I'll try out your suggestion. I still feel that it's not great for people creating the competition because the field will allow 50 characters and then we'll hit them with a warning. Was the reason you didn't want to update the 154 competitions because it would require too much effort or are there other consequences you are concerned about?

jfly commented:
I'm pretty sure updating those 154 competitions will be a hard pill to swallow for the board/results team. They're pretty adamant about not "changing history" unless there's a compelling reason. That said, if we add this warning *now*, we can always turn it into a hard validation later on (and update the 154+ competitions then).

pingskills commented:
Makes sense. I'll have a go putting back the hard 50 character limit and adding a bootstrap .alert-warning for when competition name is > 32 characters.

pedrosino commented:
One thing I just realised is that a better way to allow exceptions is using isConfirmed instead of isVisible.

Who knows how many people will see the competition between it being made visible and having the name change?
Confirmed has the desired effect (only Board or Results can change it) and doesn't run that risk.

jfly commented:
I think the problem with isConfirmed is that it would let an organizer change the competition name length to 40 just as they click the confirm button.

jfly commented on WcaOnRails/app/views/competitions/edit.html.erb:
```diff
@@ -52,9 +52,20 @@
     <% if @competition_admin_view || !is_actually_confirmed %>
       <%= f.input :id %>
+      <div class="form-group <%= !@competition.warnings[:name].nil? ? "has-warning" : "" %>">
       <%= f.input :name %>
```
nit: indentation

jfly commented on WcaOnRails/app/views/competitions/edit.html.erb:
```diff
@@ -52,9 +52,20 @@
+    <% if !@competition.warnings.empty? %>
+      <% @competition.warnings.each do |key, value| %>
+        <div class="alert alert-warning">
+        <%= key %>
```
nit: indentation

jfly commented on WcaOnRails/app/views/competitions/edit.html.erb:
```diff
@@ -52,9 +52,20 @@
+
     <% if @competition_admin_view || !is_actually_confirmed %>
       <%= f.input :id %>
+      <div class="form-group <%= !@competition.warnings[:name].nil? ? "has-warning" : "" %>">
```
Oh wow, this works? I though devise was already generating these divs for us.

Can you include a screenshot in the PR?

pingskills commented:
Hi Jeremy, Looking at the HTML you are right the form-group div is already created:

`      <div class="form-group has-warning">
        <div class="form-group string optional competition_name"><label class="string optional col-sm-2 control-label" for="competition_name">Name</label><div class="col-sm-9"><input class="string optional form-control" maxlength="50" pattern="\A(.*) (\d{4})\z" size="50" type="text" value="PingSkills Classic For Really Average Cubers 2016" name="competition[name]" id="competition_name" /><p class="help-block">The full name of the competition including the year at the end (e.g., European Rubik's Cube Championship 2006). Be sure to capitalize.</p></div></div>
      </div>`

Where does this happen and how can I add the has-warnings class to it? There are some inputs further down in the code where the form-group div is actually printed out?

The way it is at the moment just adds a redundant div but the class sticks:
![image](https://cloud.githubusercontent.com/assets/4403168/14999642/dd304496-11cd-11e6-9e71-950efe26e0e3.png)

jfly commented:
I see. The extra `<div>`s probably explain why the field looks weird. This is all generated for us by [simple_form](https://github.com/plataformatec/simple_form) (also see our simple_form config [here](https://github.com/cubing/worldcubeassociation.org/blob/master/WcaOnRails/config/initializers/simple_form_bootstrap.rb)). There might be some way to hook into it to get it to render warnings, but I feel that might be a impossible without changing the library. One solution would be to render the input by manually, rather than using simple_form. @timhabermaas, any suggestions?

pingskills commented:
I think I can use the following:

<%= f.input :name wrapper_html: { class: 'has-warning' }  %>

as described here: http://webteamalpha.com/blog/how-to-add-a-class-to-an-input-wrapper-with-simple_form/

I'll try tonight and let you know if it works.

pingskills commented:
Using <%= f.input :name, wrapper_html: { class: 'has-warning' } %> works well.

When smoke testing I noticed that _nav.html.erb already had some warning messages and info messages. So I added the warnings to the warning method in the competitions model and created a new info method for all the informational messages. I added test cases for each message as well. I think this makes it easier to add warnings and informational messages in the future and cleans things up a little.

Let me know what you think.

Here's a screenshot showing a competition with a name greater than 32 characters and and informational message about the results being uploaded soon:

![image](https://cloud.githubusercontent.com/assets/4403168/15015455/4c7e283c-124f-11e6-9d40-618c916b203f.png)

jonatanklosko commented on WcaOnRails/spec/models/competition_spec.rb:
```diff
@@ -117,18 +117,45 @@
-    competition.showAtAll = false
-    competition.name = "A really long competition name 2016"
+    competition.showAtAll = true
+    competition.name = "A really long competition name that is greater than 50 characters 2016"
```
Please make use of FactoryGirl for setting attributes:
```ruby
competition = FactoryGirl.create :competition, showAtAll: true, name: "A really long competition name that is greater than 50 characters 2016"
```
The same goes for the other tests below =)

jonatanklosko commented on WcaOnRails/spec/models/competition_spec.rb:
```diff
@@ -117,18 +117,45 @@
+  it "displays info if competition is finished but results aren't uploaded" do
+    competition = FactoryGirl.create :competition
+    competition.start_date = "2016-04-24"
+    competition.end_date = "2016-04-24"
```
Also we have a nice way for setting start/end dates, so you can write it as follows:
```ruby
competition = FactoryGirl.create :competition, starts: 1.month.ago
```
This sets the start date to the point in the past. End date will equal to the start date if you don't specify it explicitly, so you can avoid repeating.

jonatanklosko commented on WcaOnRails/spec/models/competition_spec.rb:
```diff
@@ -117,18 +117,45 @@
+  it "displays info if competition is in progress" do
+    competition = FactoryGirl.create :competition
+    competition.start_date = Date.today.to_s
+    competition.end_date = Date.today.to_s
```
The same as above:
```ruby
competition = FactoryGirl.create :competition, starts: Date.today
```

jonatanklosko commented on WcaOnRails/app/views/competitions/_nav.html.erb:
```diff
@@ -135,20 +135,20 @@
-        <% if @competition.in_progress? %>
+      <% end %>
+
+      <% if !@competition.info.empty? %>
```
More rubistic way would be to write `unless` instead of `if !something` but maybe that depends on personal feeling, so it's your choice =)

jonatanklosko commented:
Some minor comments for you, otherwise it LGTM! Expectialy the `warnings` and `info` methods are a good cleanup =)

pingskills commented:
Thanks for the feedback @jonatanklosko. 

I changed the _nav.html.erb to use unless rather than if ! as it does make it easier to read.

I've also updated the tests to make use of FactoryGirl for setting the attributes except for the one where the name is more than 50 characters as the competition doesn't seem to get created because the name fails the validation.

     Failure/Error: competition = FactoryGirl.create :competition, name: "A really long competition name that is greater than 50 characters 2016"
     ActiveRecord::RecordInvalid:
       Validation failed: Name is too long (maximum is 50 characters)

timhabermaas commented:
Try `FactoryGirl.build :competition, ...` instead of `create`. This _shouldn't_ run the validations.

jonatanklosko commented:
Actually I suggest to change all the calls of `create` to `build` here, because it doesn't create a row in the database, thus it's faster.

pingskills commented:
Thanks @timhabermaas and @jonatanklosko. I changed all of my new tests to use build.

jfly commented on WcaOnRails/app/views/competitions/_nav.html.erb:
```diff
@@ -135,20 +135,20 @@
-        <% if @competition.is_over? %>
-          <div class="alert alert-info">
-            <strong>Note:</strong> This competition is over, we are working to upload the results as soon as possible!
+      <% unless @competition.warnings.empty? %>
```
Is there any value to checking if it's empty before going into the loop? I feel like we're just going to end up calling the method twice sometimes.

pingskills replied:
Good point @jfly. I started checking a single value and then when I extended this the if stayed. It's good to take a step back after the journey and see what can be improved.

jfly commented on WcaOnRails/app/views/competitions/_nav.html.erb:
```diff
@@ -135,20 +135,20 @@
-        <% if @competition.in_progress? %>
+      <% end %>
+
+      <% unless @competition.info.empty? %>
```
Same comment as above about checking if this is empty before looping.

jfly commented on WcaOnRails/app/views/competitions/edit.html.erb:
```diff
@@ -54,7 +54,11 @@
+      <% if @competition.warnings[:name].nil? %>
+        <%= f.input :name %>
+      <% else %>
+        <%= f.input :name, wrapper_html: { class: 'has-warning' } %>
```
Nice!! Glad you figured something out here =)

Since nil is falsey, I don't think there's actually any value in if it's nil here.

I'd probably get fancy and use a ternary here. Maybe something like the following? Your call what you think is best.

```erb
<%= f.input :name, wrapper_html: { class: @competition.warnings[:name] ? "has-warning" : "" } %>
```

pingskills replied:
Very nice :)

jfly commented:
LGTM!

This is some really great cleanup, @pingskills! I have a couple of minor comments for you, but this looks so much better already.

Let me (or any member of the software team) know when you're happy with this, and one of us can merge it up for you.

pingskills commented:
Thanks @jfly. I've updated the code with your suggestions and I'm happy with this now. Thanks for everyone's help!

jfly commented:
Awesome. Great work, Jeff! I'll push it live soon.

jfly commented:
Just deployed.
